### DISCUSSION TITLE
How to get typings from query result ?

### DISCUSSION THREAD

**Discussion by mat100payette** [sqlalchemy/sqlalchemy#6687] [Usage Questions] [answered]
I have the following import and code:

```py
from sqlalchemy.engine.result import Result

def run(self, autocommit = False) -> Result:
        '''Runs the query in a distinct session and returns the result'''

        sess = session()
        result: Result = sess.execute(self.clause)
        
        if autocommit:
            sess.commit()

        return result
```
where `self.clause` is of type `ClauseElement` and `run` is a method in a `MyQuery` class.

It is used like this:
```py
rows = MyQuery(...).run().mappings().all()
results = [dict(row.items()) for row in rows]
```
---
So basically there are two things I'm wondering:

1. Is what I'm doing fine? Or in other words, is this code valid in _some_ context?
2. The `Result` class cannot be found by VSCode, but everything is fine at runtime. Therefore, I get no hints and have to deal with `Any` too often.

**Comment by CaselIT** [comment]
hi
> Is what I'm doing fine? Or in other words, is this code valid in some context?

not sure I understand what you are asking

> The Result class cannot be found by VSCode

have you installed the stub library https://github.com/sqlalchemy/sqlalchemy2-stubs ? the python plugin of python may use the old 1.3 ones by default

**Comment by zzzeek** [comment, marked as answer]
> I have the following import and code:
> 
> ```python
> from sqlalchemy.engine.result import Result
> 
> def run(self, autocommit = False) -> Result:
>         '''Runs the query in a distinct session and returns the result'''
> 
>         sess = session()
>         result: Result = sess.execute(self.clause)
>         
>         if autocommit:
>             sess.commit()
> 
>         return result
> ```
> 
> where `self.clause` is of type `ClauseElement` and `run` is a method in a `MyQuery` class.
> 
> It is used like this:
> 
> ```python
> rows = MyQuery(...).run().mappings().all()
> results = [dict(row.items()) for row in rows]
> ```
> 
> So basically there are two things I'm wondering:
> 
>     1. Is what I'm doing fine? Or in other words, is this code valid in _some_ context?

it is valid, also when you run result.mappings(), those row objects are already dictionary interfaces so typically you don't need to call dict() on them.

> 
>     2. The `Result` class cannot be found by VSCode, but everything is fine at runtime. Therefore, I get no hints and have to deal with `Any` too often.

Result is in the stubs here: https://github.com/sqlalchemy/sqlalchemy2-stubs/blob/master/sqlalchemy-stubs/engine/result.pyi#L55 

make sure you are using sqlalchemy2-stubs with vscode.

as far as typing for the objects inside the rows, we can't do that until [pep 646](https://www.python.org/dev/peps/pep-0646/) is generally available (not supported by mypy or vscode right now)

**Reply by mat100payette** [reply]
Oh, I wasn't aware of that ! How do I install it ?

**Reply by mat100payette** [reply]
Ah is there no pip package for it yet ? :o

**Reply by CaselIT** [reply]
Sure, it's called `sqlalchemy2-stubs`

**Reply by mat100payette** [reply]
Perfect thanks guys, it works great.
